src/maud/data_model/stan_input.py

```diff
@@ -0,0 +1,338 @@
+"""Definitions of inputs for Maud's Stan models."""
+
+from dataclasses import field, fields
+from math import isnan
+from typing import Dict, Sequence, Union
+
+from pydantic import StrictFloat, StrictInt, root_validator
+from pydantic.dataclasses import dataclass
+
+from maud.utils import recursively_flatten_list
+
+
+# Strict types are used here to stop pydantic from doing unwanted conversions.
+# see these links for more discussion:
+# - https://pydantic-docs.helpmanual.io/usage/models/#data-conversion
+# - https://pydantic-docs.helpmanual.io/usage/types/#strict-types
+StanDataValueBase = Union[StrictFloat, StrictInt]
+StanDataValue = Union[
+    StanDataValueBase,
+    Sequence[StanDataValueBase],
+    Sequence[Sequence[StanDataValueBase]],
+    Sequence[Sequence[Sequence[StanDataValueBase]]],
+]
+StanInputDict = Dict[str, StanDataValue]
+
+
+@dataclass
+class StanData:
+    """Something that will go in Stan's data block, in Python form."""
+
+    name: str
+    value: StanDataValue
+
+    @root_validator
+    def no_nans_allowed(cls, values):
```

Comment:
Yeah this whole section is a bit clunky, however the validation can't go in the prior because we need to validate some stan input data that isn't in a prior - for example `N_reaction`.